README.md
# BLS Keygen

![npm (tag)](https://img.shields.io/npm/v/@chainsafe/bls-keygen/latest)
[![Discord](https://img.shields.io/discord/593655374469660673.svg?label=Discord&logo=discord)](https://discord.gg/aMxzVcr)
![GitHub](https://img.shields.io/github/license/chainsafe/bls-keygen)
![ES Version](https://img.shields.io/badge/ES-2015-yellow)
![Node Version](https://img.shields.io/badge/node-10.x-green)

Utility functions for generating BLS secret keys, built for NodeJs and Browser.

- Create a master key from BIP-39 mnemonic or entropy.
- Create a derived child key from BIP-39 mnemonic, entropy, or a master key.
- Create Eth2 validator keys from a master key.

Implementation follows EIPS: [EIP-2334](https://github.com/ethereum/EIPs/pull/2334), [EIP-2333](https://github.com/ethereum/EIPs/pull/2333)

For low-level [EIP-2333](https://github.com/ethereum/EIPs/pull/2333) and [EIP-2334](https://github.com/ethereum/EIPs/pull/2334) functionality, see [@chainsafe/bls-hd-key](https://github.com/chainsafe/bls-hd-key).

### Examples
```typescript


import {
    generateRandomSecretKey,
    deriveKeyFromMnemonic,
    deriveKeyFromEntropy,
    deriveKeyFromMaster,
    deriveEth2ValidatorKeys,
} from "@chainsafe/bls-keygen";

// random secret key
const secretKey = generateRandomSecretKey();

...

// secret key from mnemonic and optional EIP-2334 path
const masterSecretKey = deriveKeyFromMnemonic(
  "impact exit example acquire drastic cement usage float mesh source private bulb twenty guitar neglect",
);

const childSecretKey = deriveKeyFromMnemonic(
  "impact exit example acquire drastic cement usage float mesh source private bulb twenty guitar neglect",
  "m/12381/3600/0/0"
);

...

// secret key from entropy and optional EIP-2334 path
const masterSecretKey = deriveKeyFromEntropy(entropy);

const childSecretKey = deriveKeyFromEntropy(
  entropy,
  "m/12381/3600/0/0"
);


...

// child secret key from master secret key and EIP-2334 path
const childSecretKey = deriveKeyFromMaster(
  masterSecretKey,
  "m/12381/3600/0/0"
);

...

// create multiple eth2 validator keys from a master secret key
const keys0 = deriveEth2ValidatorKeys(masterSecretKey, 0);
const keys1 = deriveEth2ValidatorKeys(masterSecretKey, 1);

const { signing, withdrawal } = keys0;

```

### Contribution

Requirements:
- nodejs
- yarn

```bash
    yarn install
    yarn run test
```

### Audit

This repo was audited by Least Authority as part of [this security audit](https://github.com/ChainSafe/lodestar/blob/master/audits/2020-03-23_UTILITY_LIBRARIES.pdf), released 2020-03-23. Commit [`32b068`](https://github.com/ChainSafe/bls-hd-key/commit/32b068) verified in the report.

### License

Apache-2.0


CHANGELOG.md
## 0.4.0 (2022-02-22)

### BREAKING CHANGES

* Update dependencies - Replace `Buffer` with `Uint8Array` in public APIs [#30](https://github.com/ChainSafe/bls-keygen/pull/30)

## 0.3.0 (2020-11-06)

### BREAKING CHANGES

* update bls-hd-key ([0febb0](https://github.com/chainsafe/bls-keygen/commit/0febb0))

## 0.2.1 (2020-08-11)

### Features

* Add `eth2ValidatorPaths` ([482bc6](https://github.com/chainsafe/bls-keygen/commit/482bc6))

## 0.2.0 (2020-07-30)

### BREAKING CHANGES

* update bls-hd-key ([f9f561](https://github.com/chainsafe/bls-keygen/commit/f9f561))

## 0.1.0 (2020-01-24) 

### BREAKING CHANGES

`deriveKey`
* renamed `deriveKeyFromEntropy`
* default path to `undefined` (master key)

`mnemonicToSecretKey`
* renamed `deriveKeyFromMnemonic`
* default path to `undefined` (master key)

### Features

* add `deriveKeyFromMaster` to derive a child key from a master key
* add `deriveEth2ValidatorKeys` to derive eth2 signing and withdrawal keys from a master key



test/index.spec.ts
import {deriveKeyFromMnemonic, generateRandomSecretKey, deriveKeyFromEntropy} from "../src";
import {generateMnemonic} from "@scure/bip39";
import {wordlist} from "@scure/bip39/wordlists/english";
import {expect} from "chai";

describe("random private key", function () {

  it("should generate with given entropy", function () {
    const key = generateRandomSecretKey(Buffer.alloc(32));
    const key1 = generateRandomSecretKey(Buffer.alloc(32));
    expect(key).to.not.deep.equal(key1);
  });

  it("should generate without given entropy", function () {
    const key = generateRandomSecretKey();
    const key1 = generateRandomSecretKey();
    expect(key).to.not.deep.equal(key1);
  });

});

describe("private key from mnemonic", function () {

  it("should generate master key", function () {
    const mnemonic = generateMnemonic(wordlist);
    const key = deriveKeyFromMnemonic(mnemonic);
    const key1 = deriveKeyFromMnemonic(mnemonic);
    expect(key).to.not.be.null;
    expect(key).to.deep.equal(key1);
  });

  it("should generate using given path", function () {
    const mnemonic = generateMnemonic(wordlist);
    const key = deriveKeyFromMnemonic(mnemonic, "m/12381/3600/0/1");
    const key1 = deriveKeyFromMnemonic(mnemonic, "m/12381/3600/0/2");
    expect(key).to.not.be.null;
    expect(key).to.not.deep.equal(key1);
  });

});

describe("private key from entropy", function () {

  it("should generate master key", function () {
    const seed = Buffer.alloc(32, 1);
    const key = deriveKeyFromEntropy(seed);
    const key1 = deriveKeyFromEntropy(seed);
    expect(key).to.not.be.null;
    expect(key).to.deep.equal(key1);
  });

  it("should generate child key using given path", function () {
    const seed = Buffer.alloc(32, 2);
    const key = deriveKeyFromEntropy(seed, "m/12381/3600/0/1");
    const key1 = deriveKeyFromEntropy(seed, "m/12381/3600/0/2");
    expect(key).to.not.be.null;
    expect(key).to.not.deep.equal(key1);
  });

});


test/index.test.ts
import {mnemonicToSecretKey, generateRandomSecretKey} from "../src";
import { expect } from "chai";
import {generateMnemonic} from "bip39";

describe("random private key", function () {

    it('should generate with given entropy', function () {
        const key = generateRandomSecretKey(Buffer.alloc(32)).toString("hex");
        const key1 = generateRandomSecretKey(Buffer.alloc(32)).toString("hex");
        expect(key).to.not.be.equal(key1);
    });

    it('should generate without given entropy', function () {
        const key = generateRandomSecretKey().toString("hex");
        const key1 = generateRandomSecretKey().toString("hex");
        expect(key).to.not.be.equal(key1);
    });

});

describe("private key from mnemonic", function () {

    it('should generate using default path', function () {
        const mnemonic = generateMnemonic();
       const key = mnemonicToSecretKey(mnemonic);
       const key1 = mnemonicToSecretKey(mnemonic);
       expect(key).to.not.be.null;
       expect(key.toString("hex")).to.be.equal(key1.toString("hex"));
    });

    it('should generate using given path', function () {
        const mnemonic = generateMnemonic();
       const key = mnemonicToSecretKey(mnemonic, "m/12381/60/0/1");
       const key1 = mnemonicToSecretKey(mnemonic, "m/12381/60/0/2");
       expect(key).to.not.be.null;
       expect(key.toString("hex")).to.not.be.equal(key1.toString("hex"));
    });

});

test/utils.spec.ts
import {assertValidPath, pathToIndices} from "../src/utils";
import {expect} from "chai";
import BN from "bn.js";

describe("utils", function () {

  describe("validate path", function () {

    it("should not be valid", function () {
      expect(() => assertValidPath("a/12381/60/0/0")).to.throw();
      expect(() => assertValidPath("m/12382/60/0/0")).to.throw();
      expect(() => assertValidPath("m/12381/60'/0/0")).to.throw();
      expect(() => assertValidPath("m/12381/60'/0")).to.throw();
    });

    it("should be valid", function () {
      expect(() => assertValidPath("m/12381/60/0/0")).to.not.throw();
      expect(() => assertValidPath("m/12381/90/0/0")).to.not.throw();
      expect(() => assertValidPath("m/12381/90/1/12010124124124124124124124124")).to.not.throw();
    });

  });

  describe("path to indices", function () {

    it("should convert", function () {
      const indices = pathToIndices("m/12381/60/0/1111111111111111111111111111111111111111111111111");
      expect(indices.length).to.be.equal(4);
      expect(indices[0].eqn(12381)).to.be.true;
      expect(indices[1].eqn(60)).to.be.true;
      expect(indices[2].eqn(0)).to.be.true;
      expect(indices[3].eq(new BN("1111111111111111111111111111111111111111111111111"))).to.be.true;
    });

  });

});

.eslintrc.js
const path = require('path');

module.exports = {
  root: true,
  env: {
    browser: true,
    es6: true,
    node: true,
  },
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 10,
    project: "./tsconfig.json"
  },
  plugins: [
    "@typescript-eslint",
    "eslint-plugin-import"
  ],
  extends: [
    "eslint:recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:@typescript-eslint/recommended"
  ],
  settings: {
    "import/resolver": {
      "@mpetrunic/eslint-import-resolver-lerna": {
        packages: path.resolve(__dirname, "packages")
      }
    }
  },
  rules: {
    //doesnt work, it reports false errors
    "constructor-super": "off",
    "@typescript-eslint/class-name-casing": "error",
    "@typescript-eslint/explicit-function-return-type": ["error", {
      "allowExpressions": true
    }],
    "@typescript-eslint/func-call-spacing": "error",
    "@typescript-eslint/indent": ["error", 2],
    "@typescript-eslint/interface-name-prefix": ["error", "always"],
    "@typescript-eslint/member-ordering": "error",
    "@typescript-eslint/no-explicit-any": "error",
    "@typescript-eslint/no-require-imports": "error",
    "@typescript-eslint/no-unused-vars": ["error", {
      "varsIgnorePattern": "^_"
    }],
    "@typescript-eslint/no-use-before-define": "off",
    "@typescript-eslint/semi": "error",
    "@typescript-eslint/type-annotation-spacing": "error",
    "import/no-extraneous-dependencies": ["error", {
      "devDependencies": false,
      "optionalDependencies": false,
      "peerDependencies": false
    }],
    "camelcase": "error",
    "func-call-spacing": "off",
    "max-len": ["error", {
      "code": 120
    }],
    "new-parens": "error",
    "no-caller": "error",
    "no-bitwise": "off",
    "no-conditional-assignment": true,
    "no-consecutive-blank-lines": false,
    "no-console": "warn",
    "no-var": "error",
    "object-curly-spacing": ["error", "never"],
    "object-literal-sort-keys": false,
    "prefer-const": "error",
    "quotes": ["error", "double"],
    "semi": "off"
  },
  "overrides": [
    {
      "files": ["**/test/**/*.ts"],
      "rules": {
        "import/no-extraneous-dependencies": "off",
        "@typescript-eslint/no-explicit-any": "off"
      }
    },
  ]
};

.estlintrc.js
const path = require('path');

module.exports = {
  root: true,
  env: {
    browser: true,
    es6: true,
    node: true,
  },
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 10,
    project: "./tsconfig.json"
  },
  plugins: [
    "@typescript-eslint",
    "eslint-plugin-import"
  ],
  extends: [
    "eslint:recommended",
    "plugin:import/errors",
    "plugin:import/warnings",
    "plugin:import/typescript",
    "plugin:@typescript-eslint/recommended"
  ],
  settings: {
    "import/resolver": {
      "@mpetrunic/eslint-import-resolver-lerna": {
        packages: path.resolve(__dirname, "packages")
      }
    }
  },
  rules: {
    //doesnt work, it reports false errors
    "constructor-super": "off",
    "@typescript-eslint/class-name-casing": "error",
    "@typescript-eslint/explicit-function-return-type": ["error", {
      "allowExpressions": true
    }],
    "@typescript-eslint/func-call-spacing": "error",
    "@typescript-eslint/indent": ["error", 2],
    "@typescript-eslint/interface-name-prefix": ["error", "always"],
    "@typescript-eslint/member-ordering": "error",
    "@typescript-eslint/no-explicit-any": "error",
    "@typescript-eslint/no-require-imports": "error",
    "@typescript-eslint/no-unused-vars": ["error", {
      "varsIgnorePattern": "^_"
    }],
    "@typescript-eslint/no-use-before-define": "off",
    "@typescript-eslint/semi": "error",
    "@typescript-eslint/type-annotation-spacing": "error",
    "import/no-extraneous-dependencies": ["error", {
      "devDependencies": false,
      "optionalDependencies": false,
      "peerDependencies": false
    }],
    "camelcase": "error",
    "func-call-spacing": "off",
    "max-len": ["error", {
      "code": 120
    }],
    "new-parens": "error",
    "no-caller": "error",
    "no-bitwise": "off",
    "no-conditional-assignment": true,
    "no-consecutive-blank-lines": false,
    "no-console": "warn",
    "no-var": "error",
    "object-curly-spacing": ["error", "never"],
    "object-literal-sort-keys": false,
    "prefer-const": "error",
    "quotes": ["error", "double"],
    "semi": "off"
  },
  "overrides": [
    {
      "files": ["**/test/**/*.ts"],
      "rules": {
        "import/no-extraneous-dependencies": "off",
        "@typescript-eslint/no-explicit-any": "off"
      }
    }
  ]
};

src/hdnode.ts
import {Buffer} from "buffer";
import {assertValidPath} from "./utils";
import BN from "bn.js";
import {deriveChildSK} from "./key-derivation";

export class HDNode {

    public key: Buffer;
    private depth: number;
    private path: string;

    private constructor(key: Buffer, depth: number, path: string) {
       this.key = key;
       this.depth = depth;
       this.path = path;
    }

    private derive(): HDNode {
        if(this.depth === 4) {
            return this;
        }
        const index = this.getIndex(this.path, this.depth + 1);
        return new HDNode(deriveChildSK(this.key, index), this.depth + 1, this.path).derive();
    }

    private getIndex(path: string, depth: number): BN {
        try {
            return new BN(this.path.split("/")[this.depth + 1]);
        } catch (e) {
            throw new Error(`Invalid path component #${depth} - not a number`);
        }
    }

    public static derivePath(masterKey: Buffer, path: string): Buffer {
        assertValidPath(path);
        const node = new HDNode(masterKey, 0, path);
        return node.derive().key;
    }

}

src/key-derivation.ts
import BN from "bn.js";
import {getHkdf} from "./crypto";
import SHA256 from "bcrypto/lib/sha256";
import {Buffer} from "buffer";

export function deriveMasterSK(ikm: Buffer): Buffer {
  const okm = getHkdf(ikm, 48, Buffer.from("BLS-SIG-KEYGEN-SALT-", "utf-8"));
  const okmBN = new BN(okm, "hex", "be");
  const r = new BN("52435875175126190479447740508185965837690552500527637822603658699938581184513");
  return Buffer.from(okmBN.mod(r).toArray("be", 32));
}

export function deriveChildSK(parentSK: Buffer, index: BN): Buffer {
  const compressedLamportPK = parentSKToLamportPK(parentSK, index);
  return deriveMasterSK(compressedLamportPK);
}

function parentSKToLamportPK(parentSK: Buffer, index: BN): Buffer {
  const salt = Buffer.from(index.toArray("be", 32));
  const ikm = Buffer.from(parentSK);
  const lamport0 = ikmToLamportSK(ikm, salt);
  const notIkm = Buffer.from(ikm.map((value) => ~value));
  const lamport1 = ikmToLamportSK(notIkm, salt);
  const lamportSK = lamport0.concat(lamport1);
  const lamportPK = lamportSK.map((value) => SHA256.digest(value));
  return SHA256.digest(Buffer.concat(lamportPK));
}

function ikmToLamportSK(ikm: Buffer, salt: Buffer): Buffer[] {
  const okm: Buffer = getHkdf(ikm, 8160, salt);
  return Array.from(new Array(255), (_, i) => okm.slice(i*32, (i+1)*32));
}


src/@types/bcrypto/index.d.ts
declare module "bcrypto/lib/random" {
  export function randomBytes(length: number): Buffer;
}

declare module "bcrypto/lib/sha256" {

  class SHA256 {
    native: number;
    id: string;
    size: number;
    bits: number;
    blockSize: number;
    zero: Buffer;
    ctx: SHA256;
    constructor();
    init(): this;
    update(data: Buffer): this;
    final(): Buffer;

    static hash(): SHA256;
    // static hmac(): HMAC;
    static digest(data: Buffer): Buffer;
    static root(left: Buffer, right: Buffer): Buffer;
    static multi(x: Buffer, y: Buffer, z: Buffer): Buffer;
    static mac(data: Buffer, key: Buffer): Buffer;
  }
  export = SHA256;
}

declare module "bcrypto/lib/hkdf" {

  import SHA256 = require("bcrypto/lib/sha256");

  class HKDF {

    static extract(hash: typeof SHA256, prk: Buffer, salt: Buffer): Buffer;
    static expand(hash: typeof SHA256, prk: Buffer, info: Buffer, length: number): Buffer;

  }

  export = HKDF;

}

src/index.ts
import {mnemonicToSeedSync, validateMnemonic} from "@scure/bip39";
import {wordlist} from "@scure/bip39/wordlists/english";
import {concatBytes, randomBytes} from "@noble/hashes/utils";
import {deriveChildSKMultiple, deriveMasterSK, pathToIndices} from "@chainsafe/bls-hd-key";

/**
 *
 * @param entropy optional additional entropy
 */
export function generateRandomSecretKey(entropy?: Uint8Array): Uint8Array {
  let ikm = randomBytes(32);
  if(entropy) {
    ikm = concatBytes(entropy, ikm);
  }
  return deriveKeyFromEntropy(ikm);
}

/**
 * Derive a secret key from a BIP39 mnemonic seed and optionally an EIP-2334 path.
 * If path is included, the derived key will be the child secret key at that path,
 * otherwise, the derived key will be the master secret key
 */
export function deriveKeyFromMnemonic(mnemonic: string, path?: string): Uint8Array {
  if(!validateMnemonic(mnemonic, wordlist)) {
    throw new Error("invalid mnemonic");
  }
  const ikm = Uint8Array.from(mnemonicToSeedSync(mnemonic));
  return deriveKeyFromEntropy(ikm, path);
}

/**
 * Derive a secret key from entropy and optionally an EIP-2334 path.
 * If path is included, the derived key will be the child secret key at that path,
 * otherwise, the derived key will be the master secret key
 */
export function deriveKeyFromEntropy(entropy: Uint8Array, path?: string): Uint8Array {
  const masterKey = deriveMasterSK(Uint8Array.from(entropy));
  if(path) {
    return deriveKeyFromMaster(masterKey, path);
  }
  return masterKey;
}

/**
 * Derive a child secret key from a master secret key
 * @param masterKey master secret key
 * @param path EIP-2334 path to child
 */
export function deriveKeyFromMaster(masterKey: Uint8Array, path: string): Uint8Array {
  return deriveChildSKMultiple(masterKey, pathToIndices(path));
}

export interface IEth2ValidatorKeys {
  withdrawal: Uint8Array;
  signing: Uint8Array;
}

/**
 * Return Eth2 validator HD paths
 */
export function eth2ValidatorPaths(validatorIndex: number): {
  withdrawal: string;
  signing: string;
} {
  return {
    withdrawal: `m/12381/3600/${validatorIndex}/0`,
    signing: `m/12381/3600/${validatorIndex}/0/0`,
  };
}

/**
 * Derive Eth2 validator secret keys from a single master secret key
 * @param masterKey master secret key
 */
export function deriveEth2ValidatorKeys(masterKey: Uint8Array, validatorIndex: number): IEth2ValidatorKeys {
  const paths = eth2ValidatorPaths(validatorIndex);
  return {
    withdrawal: deriveKeyFromMaster(masterKey, paths.withdrawal),
    signing: deriveKeyFromMaster(masterKey, paths.signing),
  };
}
